### DISCUSSION TITLE
Request/Ideas for Examples

### DISCUSSION THREAD

**Discussion by bastibl** [FutureSDR/FutureSDR#8] [Ideas]
We want to create some fancy examples that demonstrate what FutureSDR can do. If you have ideas, please share them with us!

- [ ] Even more portable GPU acceleration w/ [WebGPU](https://github.com/gfx-rs/wgpu), i.e., a flowgraph w/ GPU acceleration that runs on (web, Android, Linux, Windows, Mac, ...)
- [ ] Audio flowgraph (Midi input -> VST plugin (synth) -> interactive DSP (e.g., filter) -> Audio Sink) that runs native and in browser.
- [ ] FPGA acceleration example for Pluto.

**Comment by loic-fejoz** [comment]
WebGPU based NOAA images decoding?

**Reply by wcampbell0x2a** [reply]
Displaying the image in the browser would be awesome

**Comment by wcampbell0x2a** [comment]
ADS-B as an example would be good

**Comment by loic-fejoz** [comment]
And to demonstrate the parallel capability:
* [ ] Parallel decoding of RDS (Radio Data System) signals within the whole available spectrum?

**Comment by etschneider** [comment]
I've got a Pluto sitting unused in my box-of-shame. I'll to give it a go. I think the available fabric is pretty limited so I'm not sure what all we could fit. At the very least a sink and source would be nice. Some basic signal generators would be a good start.  A simple squelch block would make for a good CPU offload. Full radio rate vector source/sink for full spectrum utilization (at least in bursts).

**Reply by etschneider** [reply]
I picked up a [Xilinx Kria KR260](https://www.xilinx.com/products/som/kria/kr260-robotics-starter-kit.html). It seems like a good value for the hardware you get. 

I'll be testing FPGA acceleration on the Kria prior to the Pluto. It doesn't have a built in radio, but it does have ETH/USB3 for external radios (even an SPF+ cage if you fancy 10G).

**Reply by etschneider** [reply]
Just an FYI, the Kria is using [XRT](https://www.xilinx.com/products/design-tools/vitis/xrt.html) to configure (and interface with) the FPGA.  This is actually pretty cool as it allows for partial reconfiguration and a standard API with their accelerator platforms (and Amazon F1 instances).

It looks like they are playing with an on-fabric AXI dynamic interconnect too. Seems similar to RFNoC. 

But, it looks like I'll have to do the XRT Rust bindings myself...

(Oh, and I built FutureSDR w/ Soapy support on-target after installing the Soapy packages. Too easy.)

**Comment by JuliDi** [comment]
The spectrum example looks quite interesting. However, I feel like it lacks some documentation and seems to be limited to the HackRF. It would be cool to have this example work with other SDRs as well.

**Reply by bastibl** [reply]
Yes, I agree. It needs much more documentation. The idea was to have one flowgraph/application and show how it can work on different platforms and architectures, potentially using available accelerators.

I assume you are referring to the WASM version, which is tight to the HackRF at the moment. It would be nice to extend it but porting the drivers to WASM is non-trivial. This is happening in this repo: https://github.com/bastibl/webusb-libusb So there are some more drivers supported, which could be put in. But the actual solution would be a module similar to Soapy but for the browser. I think, this would be very valuable and would be a big push for SDR in the browser. Unfortunately, I don't have the time to work in this myself.

**Reply by JuliDi** [reply]
Interesting, I wasn't aware that you would need to port an entire driver to WASM to make that work – I somehow assumed that the drivers would just be accessible as in "normal" (i.e. non-WASM) code.

Unfortunately I can't try out the example then (lacking a HackRF), but I can put some information into a readme as a starting point for anyone interested in the topic.

**Reply by bastibl** [reply]
There are two options. Either create a WASM driver from scratch that interfaces the WebUSB API or use a shim that adapts the libusb to WebUSB and cross-compile the native driver using emscripten.

**Reply by bastibl** [reply]
If you have an RTL-SDR, it should be rather easy to change the HackRF driver to the RTL-SDR, as this is also supported by the linked repository.

**Reply by JuliDi** [reply]
Actually I got an RTL-SDR somewhere, I will give it a try!

**Reply by JuliDi** [reply]
After facing some difficulties with emscripten compiling the webusb-libusb myself, I took the files from your previous commit a92aed12a7 but keep getting either a frozen browser window or the following error(s):
![Screenshot 2022-10-22 at 12 00 22](https://user-images.githubusercontent.com/20155974/197333386-5de2ec79-ffe5-4452-b5d3-d524b76dca7b.png)

Do you have any idea what could be wrong? It still says "RTL-SDR OPEN SUCCESSFUL", but then nothing happens and the screen stays black. Using different versions of the index.html and the Rust files etc. does not make any difference. Seems to be related to the rtl-sdr driver.
I've committed the current state here, if you have the time to give it a look. https://github.com/JuliDi/FutureSDR/commit/d88d40707975d869bce4f5719384a13b20fe03b7

**Reply by bastibl** [reply]
I just created a PR in your repo with the changes that make it work for me. Very hacky stuff. It needs to be much more flashed out to make this usable.

**Reply by JuliDi** [reply]
Awesome, thank you! Now it works. Hopefully I find some time to play around with it and make it more usable.

**Reply by etschneider** [reply]
Is there a way to extend the spectrum example to allow it to process samples from a remote host? An ADALM-PLUTO for example.

FutureSDR on the Pluto could launch another FutureSDR WASM instance (via control port server) in the user's browser that communicates with the Pluto via websockets (presumably).  Would need some sort of websocket source/sink on the wasm side.

If the launching system was more beefy (e.g. a PC) it could act as a central mixer hub for all sorts of multi-party demo apps. Like the most ridiculous chat app ever - that does signal mod/demod in each user's browser (w/o need for HW radios).

**Reply by bastibl** [reply]
What is easily possible is running the CPU implementation of the spectrum example on the Pluto and the frontend on another PC. At the moment, the CPU implementation does DSP and pushes the data into a websocket. This could be changed to push raw samples (or minimally processed samples) and do more in WASM on the frontend.

Interfacing the Pluto from WASM requires more work, as the driver would have to be ported. Usually this is only libusb -> webusb but with IIO this might be more involved. I didn't have a look yet.

**Reply by etschneider** [reply]
I'm thinking that there would be separate flowgraphs running on the CPU and browser/WASM that would communicate via websocket source/sinks. The WASM fg wouldn't need to know anything about the Pluto hardware, it just communicates with the CPU fg to get the radio stream, send messages (tune), etc.

The second remote fg need not be WASM of course, it would just be neat for the Pluto (for example) to serve the secondary WASM fg itself as a self-contained demo.

**Comment by etschneider** [comment]
- [ ] NVIDIA Jetson with Vulkan accelerators

**Reply by bastibl** [reply]
I just tried on the Nano and the Xavier AGX. The included Vulkan example just works. We'd need a better idea for a real example application though.

**Reply by etschneider** [reply]
Nice! 

Perhaps VkFFT? There is a [Rust Binding](https://github.com/semio-ai/vkfft-rs).

**Reply by bastibl** [reply]
Sounds good. We could extend the vulkan version of the 'spectrum' example to do all preprocessing (FFT -> lin2db -> averaging) on the GPU. Not sure how complex this is.

**Reply by etschneider** [reply]
I think a new vulkan::d2d buffer might be helpful for that?

**Reply by bastibl** [reply]
I'm not deep enough into vulkan at the moment to understand if this really makes sense. Either one vulkan block and some shader library to compose more complex shaders inside one block. Or multiple blocks with d2d buffers. The former would be nicer but I'm not sure if there are practical limitation that I'm not aware of.

**Comment by 4xoc** [comment]
In case suggestions are still accepted, how about a simple ASK transmitter?
